File: dimediary/DimeDiaryPersistence/src/main/java/com/dimediary/model/converter/TransactionTransformer.java
package com.dimediary.model.converter;

import com.dimediary.domain.Transaction;
import com.dimediary.model.entities.BankAccountEntity;
import com.dimediary.model.entities.CategoryEntity;
import com.dimediary.model.entities.ContinuousTransactionEntity;
import com.dimediary.model.entities.TransactionEntity;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring", uses = {BankAccountTransformer.class, CategoryTransformer.class,
    ContinuousTransactionTransformer.class, UUIDTransformer.class})
public interface TransactionTransformer {

  @Mapping(source = "bankAccountEntity", target = "bankAccount")
  @Mapping(source = "continuousTransactionEntity", target = "continuousTransaction")
  @Mapping(source = "categoryEntity", target = "category")
  @Mapping(source = "transaction.amountEuroCent", target = "amountEuroCent")
  @Mapping(ignore = true, target = "timestamp")
  @Mapping(source = "transaction.id", target = "id")
  @Mapping(source = "transaction.name", target = "name")
  @Mapping(source = "transaction.fixCost", target = "fixCost")
  TransactionEntity transactionToTransactionEntity(Transaction transaction,
      BankAccountEntity bankAccountEntity,
      ContinuousTransactionEntity continuousTransactionEntity,
      CategoryEntity categoryEntity);

  Transaction transactionEntityToTransaction(
      TransactionEntity transaction);

}

File: dimediary/frontend/src/rest-services/transformer/TransactionTransformer.ts
import {Transaction} from '../../../build/openapi/models/Transaction';
import TransactionModel from '../../model/TransactionModel';
import TimeService from '@/helper/TimeService';
import {BankAccountTransformer} from "@/rest-services/transformer/BankAccountTransformer";
import {CategoryTransformer} from "@/rest-services/transformer/CategoryTransformer";
import {ContinuousTransactionTransformer} from "@/rest-services/transformer/ContinuousTransactionTransformer";
import {BankAccount} from 'build/openapi/models/BankAccount';
import {Category} from 'build/openapi/models/Category';
import {ContinuousTransaction} from 'build/openapi/models/ContinuousTransaction';


export class TransactionTransformer {

  public static to(transaction: Transaction): TransactionModel {
    return new TransactionModel(transaction.subject!,
        TimeService.isoStringToLocalDate(transaction.date!),
        transaction.amountEuroCent!, transaction.fixCost!,
        BankAccountTransformer.to(transaction.bankAccount),
        CategoryTransformer.to(transaction.category),
        ContinuousTransactionTransformer.to(transaction.continuousTransaction), transaction.id);
  }

  public static from(transactionModel: TransactionModel): Transaction {
    return new TransactionForRequest(transactionModel.id!, transactionModel.name,
        transactionModel.amountEuroCent, TimeService.localDateToIsoString(transactionModel.date!),
        BankAccountTransformer.from(transactionModel.bankAccount),
        CategoryTransformer.from(transactionModel.category),
        ContinuousTransactionTransformer.from(transactionModel.continuousTransaction),
        transactionModel.fixCost);
  }

}

class TransactionForRequest implements Transaction {


  private _id?: string;
  private _subject?: string;
  private _amountEuroCent?: number;
  private _date?: string;
  private _bankAccount?: BankAccount;
  private _category?: Category;
  private _continuousTransaction?: ContinuousTransaction;
  private _fixCost?: boolean;


  constructor(id?: string, subject?: string, amountEuroCent?: number, date?: string,
      bankAccount?: BankAccount, category?: Category, continuousTransaction?: ContinuousTransaction,
      fixCost?: boolean) {
    this._id = id;
    this._subject = subject;
    this._amountEuroCent = amountEuroCent;
    this._date = date;
    this._bankAccount = bankAccount;
    this._category = category;
    this._continuousTransaction = continuousTransaction;
    this._fixCost = fixCost;
  }


  get id(): string | undefined {
    return this._id;
  }

  get subject(): string | undefined {
    return this._subject;
  }

  get amountEuroCent(): number | undefined {
    return this._amountEuroCent;
  }

  get date(): string | undefined {
    return this._date;
  }

  get bankAccount(): BankAccount | undefined {
    return this._bankAccount;
  }

  get category(): Category | undefined {
    return this._category;
  }

  get continuousTransaction(): ContinuousTransaction | undefined {
    return this._continuousTransaction;
  }

  get fixCost(): boolean | undefined {
    return this._fixCost;
  }
}